typing/typedecl_variance.ml

```diff
@@ -88,9 +88,7 @@ let compute_variance env visited vari ty =
         compute_same ty
     | Tvar _ | Tnil | Tlink _ | Tunivar _ -> ()
     | Tpackage (_, fl) ->
-        let v =
-          Variance.(if mem Inv vari then full else unknown)
```

Comment:
In #11018 you changed the code from (1)

```ocaml
let v = Variance.(if mem Pos vari || mem Neg vari then full else unknown)
```

to (2)

```ocaml
let v = Variance.(if mem Inv vari then full else unknown)
```

and now you propose a third version (3):

```ocaml
let v = Variance.(compose vari full)
```

Why not go back to the previous version (1), is this one better? If I understand correctly, it is stronger than (1), it will return a stronger variance, while (2) was weaker than (1). Why is the new, stronger check correct?

Does the testsuite also contain negative examples checking that the variable inferred is not stronger than it should be?

Reply:
Because we now have a proper variance composition operator.
If the variance of the context is `vari` and the relative variance of a parameter is `v`, then the variance inside the parameter becomes `compose vari v`.
The previous variances were decided manually case by case, which was much more error prone.

There are tests in the testsuite in both directions (soundness and completeness), but I don't think they are exhaustive.